File: bin/Programmers/Python/완전탐색/모음사전.py


result = []
def dfs(depth,alpha,w,visited):
    global result

    result.append(w)
    
    if result == w:
        return
    
    if depth == 5:
        return
    
    for i in range(len(alpha)):
        dfs(depth + 1, alpha,w + alpha[i],visited)
        


def solution(word):
    global result
    answer = 0    
    alpha = ['A', 'E', 'I', 'O', 'U']
    visited = [False] * (len(alpha) )
    dfs(0,alpha,'',visited)
    
    answer = result.index(word)
    
    return answer



        
        
    
    
    
    

File: Programmers/Java/Level2/모음사전.java
package Programmers.Java.Level2;

import java.util.*;

public class 모음사전 {

    class Solution {
        static List<String> result = new ArrayList<>();
        static char[] alpha = {'A', 'E', 'I', 'O', 'U'};

        public int solution(String word) {
            dfs("");
            return result.indexOf(word);
        }

        public void dfs(String current) {
            result.add(current);

            if (current.length() == 5) {
                return;
            }

            for (int i = 0; i < alpha.length; i++) {
                dfs(current + alpha[i]); // 문자열은 immutable, 새로 만들어짐
            }
        }
    }

}
